lib/uploadcare/resources/resource_list.rb

```diff
@@ -0,0 +1,71 @@
+require 'ostruct'
+
+module Uploadcare
+  class Api
+    class ResourceList
+      include Enumerable
+
+      extend Forwardable
+      def_delegator :@data, :meta
+      def_delegator :@data, :objects
+      def_delegator :objects, :[]
+
+      attr_reader :options
+
+      def initialize(api, data, options)
+        @api = api
+        @data = OpenStruct.new(
+          meta: data.reject{|k, _| k == 'results'}.freeze,
+          objects: data['results'].map{|object| to_resource(api, object)}
+        )
+        @options = options.dup.freeze
+      end
+
+      def each
+        return enum_for(:each) unless block_given?
+
+        objects.each{|object| yield object}
+
+        unless last?
+          next_page = get_next_page
+          next_page.each(&Proc.new)
+        end
+
+        self
+      ensure
+        @data = OpenStruct.new(
+          meta: next_page.meta,
+          objects: objects + next_page.objects
+        ) if next_page
+      end
+
+      # TODO: #delete and #store methods
+
+      def total
+        meta['total']
+      end
+
+      def loaded
+        objects.size
+      end
+
+      private
+      attr_reader :api
+
+      def get_next_page
```

Comment:
@dmitry-mukhin no, there aren't. That's why I claim that `get_next_page` is a bad name for this method.

What `#each` method really does is it iterates through loaded objects, then asks for **the rest of the collection** (which itself is a ResourceList), calls it's each method with the same block of code and finally concatenates it's own objects with objects from this new ResourceList. 

I agree with the fact that this implementation is not straightforward. Perhaps I should rewrite it to make more understandable